### PR TITLE
Fix in-cluster API discovery documentation

### PR DESCRIPTION
**Pull request by olix0r**
The documentation incorrectly describes the way that client libraries discover the Kubernetes API server. While the `kubernetes.default.svc` DNS is provided as a convenience, **all** of the officially supported API clients use environment variables to discover the address of the API server.

This change updates the documentation to reflect this.

Fixes kubernetes/kubernetes#112263

Signed-off-by: Oliver Gould <ver@buoyant.io>

### CODE REVIEW

File: content/en/docs/tasks/run-application/access-api-from-pod.md [hunk 1]
```diff
@@ -42,10 +42,11 @@ securely with the API server.
...
 
-While running in a Pod, the Kubernetes apiserver is accessible via a Service named
-`kubernetes` in the `default` namespace. Therefore, Pods can use the
-`kubernetes.default.svc` hostname to query the API server. Official client libraries
-do this automatically.
+While running in a Pod, the Kubernetes API server is referenced by the environment
+variables `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`. As a convenience,
+the API server may also be referenced via a Service named `kubernetes` in the `default`
+namespace. Therefore, Pods can use the `kubernetes.default.svc` hostname to query the
+API server.
```
Comment (sftim) [inline content/en/docs/tasks/run-application/access-api-from-pod.md:49]: Is this true:
```suggestion
While running in a Pod, your container can work our a URL for Kubernetes API server by
fetching the two  environment variables `KUBERNETES_SERVICE_HOST` and
`KUBERNETES_SERVICE_PORT_HTTPS`.

For apps that have access to the `default` namespace, the API server may also be referenced
via a Service named `kubernetes` in the `default` namespace. This means that Pods can use 
the `kubernetes.default.svc` hostname to create a HTTPS URL for the API server.

{{< note >}}
Kubernetes does not guarantee that the API server has a valid certificate for
the hostname  `kubernetes.default.svc`;
however, the control plane **is** expected to present a valid certificate for the
hostname or IP address that `$KUBERNETES_SERVICE_HOST` represents.
{{< /note >}}
```

**Review by sftim — COMMENTED** [review]
Good spot.

I have a suggestion but I don't know if this is a claim we should make.

**Comment by sftim** [comment]
/wg api-expression

Tech review / feedback is welcome here.

File: content/en/docs/tasks/run-application/access-api-from-pod.md (hunk 1 above)
Comment (olix0r) [reply content/en/docs/tasks/run-application/access-api-from-pod.md:49]: > For apps that have access to the `default` namespace

Is this true? My understanding is that DNS discoverability isn't partitioned. So everything in the cluster should be able to find this service regardless of whether the pod has access to read services in the `default` namespace.

Comment (sftim) [reply content/en/docs/tasks/run-application/access-api-from-pod.md:49]: What I actually wanted to convey is that all clients can expect to resolve `kubernetes.default.svc`, and that additionally clients that have read access to the `kubernetes` Service (and its EndpointSlices) in NS `default` can also fetch the endpoint info another way. Probably too much detail.

Feel free to reword to suit @olix0r .

File: content/en/docs/tasks/run-application/access-api-from-pod.md [hunk 2]
```diff
@@ -42,10 +42,11 @@ securely with the API server.
 
 ### Directly accessing the REST API
 
-While running in a Pod, the Kubernetes apiserver is accessible via a Service named
-`kubernetes` in the `default` namespace. Therefore, Pods can use the
-`kubernetes.default.svc` hostname to query the API server. Official client libraries
-do this automatically.
+While running in a Pod, the Kubernetes API server is referenced by the environment
```
Comment (kbhawkey) [inline content/en/docs/tasks/run-application/access-api-from-pod.md:45]: Hi @olix0r . I like some of the suggested edits. What do you think of this update:

While running in a Pod, your container can create an HTTPS URL for the Kubernetes API server by fetching the  `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT_HTTPS` environment variables.

File: content/en/docs/tasks/run-application/access-api-from-pod.md (hunk 1 above)
Comment (olix0r) [reply content/en/docs/tasks/run-application/access-api-from-pod.md]: Rereading your suggestion, I see what you mean now.

Practically, it seems unlikely that a pod that already has access to the API server will need to read the Service/Endpoints--this information is almost always used when bootstrapping a client that does not yet have API access.

I've  kept your note about the TLS certificate, however.

**Comment by tengqm** [comment]
What's the next step for this PR?
Shall we merge it as is or continue waiting for some tech LGTM?

**Comment by olix0r** [comment]
@sftim @kbhawkey Are you happy with the changes?

**Comment by sftim** [comment]
Looks good to me.

/lgtm

**Comment by tengqm** [comment]
@olix0r The gate is not happy with the change yet. Please try recommit this PR so that another test can be triggered?

**Comment by divya-mohan0209** [comment]
Hello @olix0r , please can you recommit this PR so that another test can be triggered?

**Comment by sftim** [comment]
I can trigger a build for you @olix0r

**Comment by olix0r** [comment]
Sorry about that. I've merged main into my branch.

**Review by sftim — COMMENTED** [review]
Let's ship it.

/label tide/merge-method-squash
/lgtm
/approve

**Pull request merged**